File: Data Structures & Algorithms/products-of-array-discluding-self/submission-3.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        ans = [0] * len(nums)
        left = [1] * len(nums)
        left[0] = nums[0];
        right = [1] * len(nums)
        right[len(nums)-1] = nums[len(nums)-1]
        
        for i in range( 1,len(nums),1):
            left[i] = left[i-1] * nums[i]
        
        for i in range(len(nums) -2,-1,-1):
            right[i] = right[i+1]* nums[i]
        
        print(left)
        print(right)
        
        for i in range(0,len(ans)):
            if i == 0:
                ans[i] = right[1]
            elif i == len(nums)-1:
                ans[i] = left[len(nums)-2]
            else:
                ans[i] = left[i-1] * right[i+1]
        return ans

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    static int ans;
    public int maxPathSum(TreeNode root) {
        ans = Integer.MIN_VALUE;;
        findMaxSum(root);
        return ans;
    }

    public static int findMaxSum(TreeNode root){
        if(root == null){
            return 0;
        }
        int leftSum = findMaxSum(root.left);
        int rightSum = findMaxSum(root.right);
        int maxSum = Math.max(leftSum, rightSum);
        if(maxSum + root.val>ans){
            ans = maxSum + root.val;
        }
        if(leftSum + rightSum + root.val> ans){
            ans = leftSum + rightSum + root.val;
        }
        if(root.val>ans){
         ans = root.val;
        }
        return Math.max(maxSum + root.val, root.val);
    }
}
